src/auth/backchannel.ts

```diff
@@ -0,0 +1,254 @@
+// Wednesday, 8 January, 2025
+// Client Initiated Backchannel Authentication (CIBA)
+
+// CIBA is an OpenID Foundation standard for a decoupled authentication flow. It enables
+// solution developers to build authentication flows where the user logging in does not do so
+// directly on the device that receives the ID or access tokens (the “Consumption Device”), but
+// instead on a separate “Authorization Device”.
+
+import { JSONApiResponse } from '../lib/models.js';
+import { BaseAuthAPI } from './base-auth-api.js';
+
+/**
+ * The response from the authorize endpoint.
+ */
+export type AuthorizeResponse = {
+  /**
+   * The authorization request ID.
+   */
+  auth_req_id: string;
+  /**
+   * The duration in seconds until the authentication request expires.
+   */
+  expires_in: number;
+  /**
+   * The interval in seconds to wait between poll requests.
+   */
+  interval: number;
+};
+
+type AuthorizeCredentialsPartial = {
+  client_id: string;
+  client_secret?: string;
+  client_assertion?: string;
+  client_assertion_type?: string;
+};
+
+/**
+ * The login hint containing information about the user for authentication.
+ */
+type LoginHint = {
+  /**
+   * The format of the login hint.
+   */
+  format: 'iss_sub';
+  /**
+   * The issuer URL.
+   */
+  iss: string;
+  /**
+   * The subject identifier.
+   */
+  sub: string;
+};
+
+/**
+ * Generates the login hint for the user.
+ *
+ * @param {string} userId - The user ID.
+ * @param {string} domain - The tenant domain.
+ * @returns {string} - The login hint as a JSON string.
+ */
+const getLoginHint = (userId: string, domain: string): string => {
+  // remove trailing '/' from domain, added later for uniformity
+  const trimmedDomain = domain.endsWith('/') ? domain.slice(0, -1) : domain;
+  const loginHint: LoginHint = {
+    format: 'iss_sub',
+    iss: `https://${trimmedDomain}/`,
+    sub: `${userId}`,
+  };
+  return JSON.stringify(loginHint);
+};
+
+/**
+ * Options for the authorize request.
+ */
+export type AuthorizeOptions = {
+  /**
+   * A human-readable string intended to be displayed on both the device calling /bc-authorize and the user’s authentication device.
+   */
+  binding_message: string;
+  /**
+   * A space-separated list of OIDC and custom API scopes.
+   */
+  scope: string;
+  /**
+   * Unique identifier of the audience for an issued token.
+   */
+  audience?: string;
+  /**
+   * Custom expiry time in seconds for this request.
+   */
+  request_expiry?: string;
+  /**
+   * The user ID.
+   */
+  userId: string;
```

Comment:
@tusharpandey13 What's the reason for using camel-case here when the rest of the options are in snake-case?